### PR TITLE
Add Metal support

### LINKED ISSUE #326: Add Metal support
As the M1 GPUs are becoming popular, it would be great to add Metal support to KernelAbstractions.jl. 

cc @maleadt

### PR DESCRIPTION
**Pull request by tgymnich**
fixes #326

### CODE REVIEW

**Comment by vchuravy** [comment]
Depends on https://github.com/JuliaGPU/Metal.jl/pull/61

@maleadt is there a way to get CI?

**Comment by maleadt** [comment]
> @maleadt is there a way to get CI?

Yes, but I couldn't figure out sandboxing, so it's behind `bors`.

**Comment by maleadt** [comment]
> @maleadt is there a way to get CI?

Yes, using `bors`.

File: .ci/develop.jl [hunk 1]
```diff
@@ -28,6 +28,11 @@ if !CI || BACKEND == "oneAPI"
     push!(pkgs, oneapikernels)
 end
 
+if !CI || BACKEND == "Metal"
+    metalkernels = Pkg.PackageSpec(path = joinpath(root_directory, "lib", "MetalKernels"))
+    push!(pkgs, metalkernels)
```
Comment (vchuravy) [inline .ci/develop.jl:33]:
```suggestion
    push!(pkgs, metalkernels)
    push!(pkgs, PackageSpec(name="Metal", rev="main")) # TODO remove when tagged
```

Comment (tgymnich) [reply .ci/develop.jl:33]: thanks

File: .buildkite/pipeline.yml [hunk 2]
```diff
@@ -96,6 +96,26 @@ steps:
...
+    agents:
+      queue: "juliaecosystem"
+      os: "macos"
+      arch: "aarch64"
+    env:
+      KERNELABSTRACTIONS_TEST_BACKEND: "Metal"
+    timeout_in_minutes: 60
 
 env:
   JULIA_PKG_SERVER: "" # it often struggles with our large artifacts
```
Comment (vchuravy) [inline .buildkite/pipeline.yml:121]:
```suggestion
```

File: .ci/develop.jl [hunk 3]
```diff
@@ -28,6 +28,12 @@ if !CI || BACKEND == "oneAPI"
...
+    push!(pkgs, PackageSpec(name="Metal", rev="main")) # TODO remove when tagged
```
Comment (vchuravy) [inline .ci/develop.jl:34]:
```suggestion
    push!(pkgs, Pkg.PackageSpec(name="Metal", rev="main")) # TODO remove when tagged
```

File: .ci/develop.jl [hunk 4]
```diff
@@ -28,6 +28,12 @@ if !CI || BACKEND == "oneAPI"
     push!(pkgs, oneapikernels)
 end
 
+if !CI || BACKEND == "Metal"
+    metalkernels = Pkg.PackageSpec(path = joinpath(root_directory, "lib", "MetalKernels"))
+    push!(pkgs, metalkernels)
+    push!(pkgs, Pkg.PackageSpec(name="Metal", rev="main")) # TODO remove when tagged
```
Comment (vchuravy) [inline .ci/develop.jl:34]:
```suggestion
    Pkg.add(Pkg.PackageSpec(name="Metal", rev="main")) # TODO remove when tagged
```

**Comment by vchuravy** [comment]
@tgymnich can disable the offending tests for Metal so that we can merge this?

**Comment by tgymnich** [comment]
@vchuravy we could. But I would need to take a second look at the locking logic before we merge.

**Comment by vchuravy** [comment]
Seqfault on nightly and a error in `async_copy` for 1.8

**Comment by maleadt** [comment]
I'll be releasing Metal.jl soon, so this should be able to move forward then.

**Comment by tgymnich** [comment]
Since there seems to be something wrong with `wait` I implemented my own method waiting for a GPU event in objc.

**Comment by maleadt** [comment]
That seems like something we should add to Metal.jl. That said, `wait_completed` shouldn't hang, so mind filing an issue if you can reproduce?

**Pull request merged**